R/checkModels.R

```diff
@@ -1,74 +1,201 @@
+    person.groups$group <- all.persons.lab
+    all.persons <- FALSE
+  }
+  # person.groups auf Plausibilitaet checken
+  if(!is.null(person.groups)){
+    # hat erste Spalte mehr Elemente als alle anderen
+    len <- sapply(person.groups, function(x) length(unique(x)))
+    len.log <- len < len[1]
+    if(!all(len.log[-1])) warning(paste0("first column of person.groups might not contain person ids; please check\n(number of unique elements is smaller than in another column)"),
+                                  call. = FALSE)
+  }
+
+  return(d)
+}
+
+### called by defineModel() and splitModels() ----------------------------------
+
+checkQmatrixConsistency <- function(qmat){
+  qmat <- eatTools::makeDataFrame(qmat, name = "Q matrix")
+  if(!inherits(qmat[,1], "character")){
+    qmat[,1] <- as.character(qmat[,1])}
+
+  nClass<- sapply(qmat[,-1,drop=FALSE], inherits, what=c("numeric", "integer"))
+  if(!all(nClass)){
+    warning(paste0("Found non-numeric indicator column(s) in the Q matrix. Transform column(s) '",
+                   paste(colnames(qmat)[ which(nClass==FALSE)+1], collapse = "', '") ,"' into numeric format."))
+    qmat <- data.frame(qmat[,1,drop=FALSE], eatTools::asNumericIfPossible(qmat[,-1,drop=FALSE]), stringsAsFactors = FALSE)}
+
+  werte <- eatTools::tableUnlist(qmat[,-1,drop=FALSE], useNA="always")
+  if(length(setdiff(names(werte), c("0","1", "NA")))<0){
+    stop("Q matrix must not contain entries except '0' and '1'.\n")}
+  if(werte[match("NA", names(werte))] > 0){
+    stop("Missing values in Q matrix.\n")}
+
+  wertes <- lapply(qmat[,-1,drop=FALSE], FUN = function(col) {all(col == 0)})
+  konst <- which(wertes == TRUE)
+  if(length(konst)>0){
+    cat(paste0("Column(s) '",paste(names(konst), collapse = "', '"),"' in Q matrix are constant with value 0. Delete column(s).\n"))
+    qmat <- qmat[,-match(names(konst), colnames(qmat)), drop=FALSE]
+  }
+
+  doppel <- which(duplicated(qmat[,1]))
+  if(length(doppel)>0){
+    cat("Found duplicated elements in the item id column of the q matrix. Duplicated elements will be removed.\n")
+    chk  <- table(qmat[,1])
+    chk  <- chk[which(chk > 1)]
+    chkL <- lapply(names(chk), FUN = function(ch){
+      qChk <- qmat[which(qmat[,1] == ch),]
+      pste <- apply(qChk, 1, FUN = function(x) {paste(x[-1], collapse="")})
+      if(!all(pste == pste[1])){
+        stop("Inconsistent q matrix.\n")}
+    })
+    qmat <- qmat[!duplicated(qmat[,1]),]
+  }
+
+  zeilen <- apply(qmat, 1, FUN = function(y) {all(names(table(y[-1])) == "0")} )
+  weg    <- which(zeilen == TRUE)
+  if(length(weg)>0) {
+    cat(paste("Note: Following ",length(weg)," item(s) in Q matrix do not belong to any dimension. Delete these item(s) from Q matrix.\n",sep=""))
+    cat("    "); cat(paste(qmat[weg,1],collapse=", ")); cat("\n")
+    qmat <- qmat[-weg,]
+  }
+
+  # wenn nur eine Spalte wird diese als IDs angenommen
```

Comment:
Lines 89ff: 
```r
# wenn nur eine Spalte wird diese als IDs angenommen
if(test_data_frame(qmat, ncols = 1)){
    warning("qMatrix contains just one column; this is treated as item names", call. = FALSE)
    qMatrix$dim <- 1
  }
```
That could be problematic. I don't know if you can use these checks for the functions that belong to `defineModel()`. I think it would be better if the function gives an error if the Q matrix only has one column.

Reply:
I took a closer look at this. `defineModel()` and `splitModel()` both require a Q matrix if necessary. The consistency checks are similar, but not exactly the same. For example, the Q matrix for `defineModel()` may also contain numerical values beyond 0 and 1 (even if this is only required in extremely rare cases), but in the Q matrix for `splitModels()` they must always be exclusively 0 or 1. I have therefore adapted the `checkQmatrixConsistency()` function slightly so that you can parameterize whether only a warning (`defineModel()`) or an error (`splitModels()`) is given for values other than 0 or 1. You can then use this check function for both variants. There was also a small error in the function that occurred when users specify a two-dimensional model (Q matrix with two indicator columns), but there are only items of one dimension in the data set. This can lead to a situation where, after excluding the items that are in the Q matrix but not in the data set, one column of the Q matrix contains only 0 values. To avoid merge conflicts, I have made the adjustments here:
```
t:\Sebastian\eatModel\checkqmatixconsistency.r
t:\Sebastian\eatModel\splitModels.R
```
You can see if you can follow along and, if so, apply the changes to your commit?

Reply:
I applied the changes from `checkQmatrixConsistency`.